### PR TITLE
Improve relay code and prepare for relay management API

### LINKED ISSUE #169: HoprConnect seems to assign twice to relay nodes
For some reason, HoprConnect is assigning twice to the same relay node.

Example debug output:

```
  hopr-core /p2p/16Uiu2HAmQESutWDjkmBsSZYtVnd3zACCUfR2Bmy4rpnsqDBa6ECY/p2p-circuit/p2p/16Uiu2HAm6gkgwMKBs4X63mvfBoWPqvgFtxuHUTWbBxeu5fCu7UC2 +0ms
  hopr-core /p2p/16Uiu2HAmBH1WR5wvuQEGq6gQ6BKDFU57U5uefGBrqq4RBedDsCXt/p2p-circuit/p2p/16Uiu2HAm6gkgwMKBs4X63mvfBoWPqvgFtxuHUTWbBxeu5fCu7UC2 +1ms
```

### PR DESCRIPTION
**Pull request by robertkiel**
**Rationale:**
- Cleanup and decapsulate relay code from rest
- Manage state and test state management
- Improve testing

**Breaking changes:**
- errors became numbers, e.g. `FAIL_COULD_NOT_REACH_COUNTERPARTY` became `Uint8Array([2])`
- length-prefixed encoding in WebRTC
- delete record once relayed connection once connection is torn down
- way stricter unit testing

**Changes:**
- fixed various deadlocks in connection establishment
- depromisification where possible
- switch from integration tests to fine-granular unit tests, nearly complete rewrite of existing test suite
- seperated relay state management code from relay code
- seperated relay handshake sub-protocol from relay code
- moved entire relay code from top-level `index.ts` into `relay/index.ts`
- refactor of repo structure: `base`, `relay`, `webrtc` to make clear which functionality is implemented in which file, e.g. `webRTCConnection.ts` -> `webrtc/connection.ts` and `relayConnection.ts` -> `relay/connection.ts`
- priority-ordering of status messages
- length encode messages sent through WebRTC to make sure merged messages get properly decoded
- various event handlers removed once connected / errored

Fixes #169 , #163 , #64 , #61

**Testing:**

Unit test:
`yarn test` and preferably with `DEBUG=hopr*,simple-peer yarn test`

Integration test:
see https://github.com/hoprnet/hopr-connect/blob/main/TESTING.md
- spawn a bootstrap server, acting as STUN and TURN server
- spawn two nodes and use bootstrap server for STUN and TURN

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -152,6 +156,10 @@ class HoprConnect implements Transport {
     )
   }
 
+  addRelay() {}
```
Comment (tolbrino) [inline src/index.ts:159]: Can you document these placeholders?

**Review by tolbrino — COMMENTED** [review]
There is no inline documentation. A change of this size should take the opportunity and add some info and context.

**Comment by tolbrino** [comment]
unit and integration tests are working as expected

File: src/index.ts (hunk 1 above)
Comment (robertkiel) [reply src/index.ts:163]: done

**Review by tolbrino — APPROVED** [review]

**Pull request merged**